File: solver.mjs
const CX = [0, 1, 0, -1, 0, 1, -1, -1, 1];
const CY = [0, 0, 1, 0, -1, 1, 1, -1, -1];
const OPP = [0, 3, 4, 1, 2, 7, 8, 5, 6];
const W = [4 / 9, 1 / 9, 1 / 9, 1 / 9, 1 / 9, 1 / 36, 1 / 36, 1 / 36, 1 / 36];
export const SHAPES = ['airfoil', 'venturi', 'coanda', 'cylinder', 'plate', 'empty'];
export function speedToKmh(speed, metres, seconds) {
  if (![speed, metres, seconds].every(Number.isFinite) || metres <= 0 || seconds <= 0) throw new Error('Physical scales must be positive and finite.');
  return speed * metres / seconds * 3.6;
}
export function bodyAt(X, Y, shape, angle) {
  const a = angle * Math.PI / 180, dx = X - 0.72, dy = Y - 0.5;
  const rx = dx * Math.cos(a) - dy * Math.sin(a), ry = dx * Math.sin(a) + dy * Math.cos(a);
  if (shape === 'cylinder') return dx * dx + dy * dy < 0.125 ** 2;
  if (shape === 'plate') return Math.abs(rx) < 0.16 && Math.abs(ry) < 0.018;
  if (shape === 'airfoil') {
    const t = (rx + 0.22) / 0.44;
    return t >= 0 && t <= 1 && Math.abs(ry) <= 5 * 0.18 * 0.44 * (0.2969 * Math.sqrt(t) - 0.126 * t - 0.3516 * t ** 2 + 0.2843 * t ** 3 - 0.1036 * t ** 4);
  }
  if (shape === 'venturi') {
    const neck = Math.abs(X - 0.95) < 0.65 ? 0.5 * (1 + Math.cos(Math.PI * (X - 0.95) / 0.65)) : 0;
    return Y < 0.08 + 0.23 * neck || Y > 0.92 - 0.23 * neck;
  }
  return shape === 'coanda' && ((X < 0.65 && Y >= 0.4) || ((X - 0.65) ** 2 + (Y - 0.65) ** 2 < 0.25 ** 2));
}
export function equilibrium(q, rho, u, v) {
  const cu = CX[q] * u + CY[q] * v;
  return W[q] * rho * (1 + 3 * cu + 4.5 * cu * cu - 1.5 * (u * u + v * v));
}
export class Solver {
  constructor({ width = 288, shape = 'airfoil', speed = 0.035, viscosity = 0.04, angle = 5, heat = 60, source = 'tunnel', fanX = 0.06, fanY = 0.5, direction = 0 } = {}) {
    if (![192, 288, 384, 576].includes(width) || !SHAPES.includes(shape) || !['fan', 'tunnel'].includes(source) || ![speed, viscosity, angle, heat, fanX, fanY, direction].every(Number.isFinite) || speed < 0 || speed > 0.09 || viscosity < 0.02 || viscosity > 0.12 || Math.abs(angle) > 20 || heat < 0 || heat > 100 || fanX < 0.03 || fanX > 0.97 || fanY < 0.08 || fanY > 0.92 || Math.abs(direction) > 180) throw new Error('Invalid simulation parameters.');
    Object.assign(this, { width, height: width / 2, shape, speed, viscosity, angle, heat, source, fanX, fanY, direction, steps: 0 });
    const n = this.n = width * this.height;
    this.f = new Float32Array(n * 9); this.next = new Float32Array(n * 9);
    this.rho = new Float32Array(n).fill(1); this.u = new Float32Array(n); this.v = new Float32Array(n);
    this.temperature = new Float32Array(n).fill(20); this.nextTemperature = new Float32Array(n);
    this.solid = new Uint8Array(n); this.wall = new Uint8Array(n);
    const h = this.height;
    for (let y = 0; y < h; y++) for (let x = 0; x < width; x++) {
      const i = x + y * width, body = bodyAt(x / h, y / h, shape, angle);
      this.wall[i] = y === 0 || y === h - 1 ? 1 : 0;
      this.solid[i] = body || this.wall[i] ? 1 : 0;
      if (body && !this.wall[i]) this.temperature[i] = heat;
      const u = this.solid[i] ? 0 : this.inlet(y); this.u[i] = u;
      for (let q = 0; q < 9; q++) this.f[i * 9 + q] = equilibrium(q, 1, u, 0);
    }
    this.buildFan();
  }
  buildFan() {
    this.fan = [];
    const a = this.direction * Math.PI / 180, c = Math.cos(a), s = Math.sin(a), h = this.height;
    this.fanU = c * this.speed; this.fanV = s * this.speed;
    if (this.source !== 'fan') return;
    for (let y = 1; y < h - 1; y++) for (let x = 1; x < this.width - 1; x++) {
      const dx = x - this.fanX * (this.width - 1), dy = y - this.fanY * (h - 1);
      const along = (dx * c + dy * s) / (h * 0.025), across = (-dx * s + dy * c) / (h * 0.09);
      if (Math.abs(along) < 1 && Math.abs(across) < 1 && !this.solid[x + y * this.width]) this.fan.push([x + y * this.width, 0.15 * (1 - along * along) * (1 - across * across)]);
    }
  }
  update(config) {
    // Validate the whole request before mutating the live field.
    const candidate = new Solver(config);
    if (candidate.width !== this.width || candidate.shape !== this.shape) throw new Error('Shape and resolution changes require reset.');
    const oldSolid = this.solid;
    // ponytail: re-equilibrate only changed cells; use conservative moving boundaries for quantitative pitching-wing loads.
    for (let i = 0; i < this.n; i++) {
      if (oldSolid[i] !== candidate.solid[i]) {
        let r = 0, ux = 0, vy = 0, temperature = 0, count = 0;
        for (const j of [i - 1, i + 1, i - this.width, i + this.width]) if (j >= 0 && j < this.n && !oldSolid[j] && !candidate.solid[j]) {
          r += this.rho[j]; ux += this.u[j]; vy += this.v[j]; temperature += this.temperature[j]; count++;
        }
        r = count ? r / count : 1; ux = count ? ux / count : 0; vy = count ? vy / count : 0;
        if (candidate.solid[i]) { ux = 0; vy = 0; }
        this.rho[i] = r; this.u[i] = ux; this.v[i] = vy;
        this.temperature[i] = count ? temperature / count : 20;
        for (let q = 0; q < 9; q++) this.f[i * 9 + q] = this.next[i * 9 + q] = equilibrium(q, r, ux, vy);
      }
      if (candidate.solid[i]) this.temperature[i] = candidate.wall[i] ? 20 : candidate.heat;
    }
    for (const key of ['speed', 'viscosity', 'angle', 'heat', 'source', 'fanX', 'fanY', 'direction', 'solid', 'fan', 'fanU', 'fanV']) this[key] = candidate[key];
  }
  inlet(y) {
    if (this.source === 'fan') return 0;
    if (this.shape !== 'coanda') return this.speed;
    const t = (y / this.height - 0.32) / 0.08;
    return t > 0 && t < 1 ? this.speed * 4 * t * (1 - t) : 0;
  }
  step() {
    const { width: w, height: h, n, f, next, solid, rho, u, v } = this;
    const omega = 1 / (0.5 + 3 * this.viscosity);
    // Local momentum actuator: relax velocity, preserving cell density (no mass injection).
    for (const [i, strength] of this.fan) {
      const ux = u[i] + strength * (this.fanU - u[i]), vy = v[i] + strength * (this.fanV - v[i]);
      for (let q = 0; q < 9; q++) f[i * 9 + q] += equilibrium(q, rho[i], ux, vy) - equilibrium(q, rho[i], u[i], v[i]);
      u[i] = ux; v[i] = vy;
    }
    // BGK collision, push streaming, halfway no-slip bounce-back.
    for (let i = 0; i < n; i++) {
      if (solid[i]) continue;
      const x = i % w, y = (i / w) | 0, base = i * 9;
      for (let q = 0; q < 9; q++) {
        const value = f[base + q] + omega * (equilibrium(q, rho[i], u[i], v[i]) - f[base + q]);
        const nx = x + CX[q], ny = y + CY[q], j = nx + ny * w;
        if (ny < 0 || ny >= h || (nx >= 0 && nx < w && solid[j])) next[base + OPP[q]] = value;
        else if (nx >= 0 && nx < w) next[j * 9 + q] = value;
      }
    }
    for (let y = 1; y < h - 1; y++) {
      const i = y * w, b = i * 9;
      if (!solid[i] && this.source === 'fan') {
        for (let q = 0; q < 9; q++) next[b + q] = next[b + 9 + q];
      } else if (!solid[i]) {
        // Zou/He inlet reconstructs the three incoming populations.
        const ux = this.inlet(y);
        const r = (next[b] + next[b + 2] + next[b + 4] + 2 * (next[b + 3] + next[b + 6] + next[b + 7])) / (1 - ux);
        next[b + 1] = next[b + 3] + 2 * r * ux / 3;
        next[b + 5] = next[b + 7] + (next[b + 4] - next[b + 2]) / 2 + r * ux / 6;
        next[b + 8] = next[b + 6] + (next[b + 2] - next[b + 4]) / 2 + r * ux / 6;
      }
      const out = i + w - 1;
      if (!solid[out]) for (let q = 0; q < 9; q++) next[out * 9 + q] = next[(out - 1) * 9 + q];
    }
    this.f = next; this.next = f;
    for (let i = 0; i < n; i++) {
      if (solid[i]) continue;
      let r = 0, ux = 0, vy = 0;
      for (let q = 0; q < 9; q++) { const value = next[i * 9 + q]; r += value; ux += value * CX[q]; vy += value * CY[q]; }
      ux /= r; vy /= r;
      if (!Number.isFinite(r + ux + vy) || r < 0.5 || r > 1.5 || ux * ux + vy * vy > 0.04) throw new Error('Low-Mach stability limit exceeded. Reduce speed or increase viscosity, then reset.');
      rho[i] = r; u[i] = ux; v[i] = vy;
    }
    // ponytail: first-order upwind heat transport is diffusive; upgrade for quantitative heat-transfer studies.
    const T = this.temperature, out = this.nextTemperature, alpha = this.viscosity / 0.71;
    for (let y = 0; y < h; y++) for (let x = 0; x < w; x++) {
      const i = x + y * w;
      if (solid[i]) { out[i] = this.wall[i] ? 20 : this.heat; continue; }
      if (x === 0) { out[i] = 20; continue; }
      const t = T[i], left = T[i - 1], right = T[x === w - 1 ? i : i + 1], top = T[i - w], bottom = T[i + w];
      out[i] = t - u[i] * (u[i] >= 0 ? t - left : right - t) - v[i] * (v[i] >= 0 ? t - top : bottom - t) + alpha * (left + right + top + bottom - 4 * t);
    }
    this.temperature = out; this.nextTemperature = T; this.steps++;
  }
  snapshot() {
    let reference = 0, count = 0, maxSpeed = 0;
    for (let y = 1; y < this.height - 1; y++) { const i = (y + 1) * this.width - 1; if (!this.solid[i]) { reference += this.rho[i]; count++; } }
    reference /= count || 1;
    const fields = new Float32Array(this.n * 4);
    for (let i = 0; i < this.n; i++) {
      fields[i * 4] = this.u[i]; fields[i * 4 + 1] = this.v[i]; fields[i * 4 + 2] = this.solid[i] ? 0 : (this.rho[i] - reference) / 3; fields[i * 4 + 3] = this.temperature[i];
      maxSpeed = Math.max(maxSpeed, Math.hypot(this.u[i], this.v[i]));
    }
    return { fields, steps: this.steps, maxSpeed, reynolds: this.speed * (this.height / 4) / this.viscosity };
  }
}

File: solver.cpp
#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>
#include <emscripten/bind.h>

using emscripten::val;
constexpr int CX[] = {0, 1, 0, -1, 0, 1, -1, -1, 1};
constexpr int CY[] = {0, 0, 1, 0, -1, 1, 1, -1, -1};
constexpr int OPP[] = {0, 3, 4, 1, 2, 7, 8, 5, 6};
constexpr double W[] = {4./9, 1./9, 1./9, 1./9, 1./9, 1./36, 1./36, 1./36, 1./36};
constexpr double PI = 3.14159265358979323846;
double equilibrium(int q, double rho, double u, double v) {
  if (q < 0 || q > 8) throw std::runtime_error("Invalid lattice direction.");
  const double cu = CX[q] * u + CY[q] * v;
  return W[q] * rho * (1 + 3 * cu + 4.5 * cu * cu - 1.5 * (u * u + v * v));
}
bool bodyAt(double X, double Y, int shape, double angle) {
  const double a = angle * PI / 180, dx = X - .72, dy = Y - .5;
  const double rx = dx * std::cos(a) - dy * std::sin(a), ry = dx * std::sin(a) + dy * std::cos(a);
  if (shape == 3) return dx * dx + dy * dy < .125 * .125;
  if (shape == 4) return std::abs(rx) < .16 && std::abs(ry) < .018;
  if (shape == 0) {
    const double t = (rx + .22) / .44;
    return t >= 0 && t <= 1 && std::abs(ry) <= 5 * .18 * .44 * (.2969 * std::sqrt(t) - .126 * t - .3516 * t*t + .2843 * t*t*t - .1036 * t*t*t*t);
  }
  if (shape == 1) {
    const double neck = std::abs(X - .95) < .65 ? .5 * (1 + std::cos(PI * (X - .95) / .65)) : 0;
    return Y < .08 + .23 * neck || Y > .92 - .23 * neck;
  }
  return shape == 2 && ((X < .65 && Y >= .4) || ((X - .65)*(X - .65) + (Y - .65)*(Y - .65) < .25*.25));
}
struct Config {
  int width, shape, source;
  double speed, viscosity, angle, heat, fanX, fanY, direction;
  void validate() const {
    if ((width != 192 && width != 288 && width != 384 && width != 576) || shape < 0 || shape > 5 || source < 0 || source > 1)
      throw std::runtime_error("Invalid simulation parameters.");
    for (double x : {speed, viscosity, angle, heat, fanX, fanY, direction})
      if (!std::isfinite(x)) throw std::runtime_error("Simulation parameters must be finite.");
    if (speed < 0 || speed > .09 || viscosity < .02 || viscosity > .12 || std::abs(angle) > 20 || heat < 0 || heat > 100 || fanX < .03 || fanX > .97 || fanY < .08 || fanY > .92 || std::abs(direction) > 180)
      throw std::runtime_error("Invalid simulation parameters.");
  }
};
class Solver {
  Config c;
  int h, n;
  double steps = 0, fanU = 0, fanV = 0, maxSpeed = 0;
  bool failed = false;
  std::vector<double> f, next, rho, u, v, temperature, nextTemperature, fields;
  std::vector<unsigned char> solid, wall;
  std::vector<std::pair<int, double>> fan;
  void buildFan() {
    fan.clear();
    const double a = c.direction * PI / 180, co = std::cos(a), si = std::sin(a);
    fanU = co * c.speed; fanV = si * c.speed;
    if (!c.source) return;
    for (int y = 1; y < h - 1; ++y) for (int x = 1; x < c.width - 1; ++x) {
      const double dx = x - c.fanX * (c.width - 1), dy = y - c.fanY * (h - 1);
      const double along = (dx * co + dy * si) / (h * .025), across = (-dx * si + dy * co) / (h * .09);
      if (std::abs(along) < 1 && std::abs(across) < 1 && !solid[x + y * c.width])
        fan.emplace_back(x + y * c.width, .15 * (1 - along*along) * (1 - across*across));
    }
  }
public:
  explicit Solver(Config config) : c(config) {
    c.validate(); h = c.width / 2; n = c.width * h;
    f.resize(n*9); next.resize(n*9); rho.assign(n, 1); u.resize(n); v.resize(n);
    temperature.assign(n, 20); nextTemperature.resize(n); fields.resize(n*4); solid.resize(n); wall.resize(n);
    for (int y = 0; y < h; ++y) for (int x = 0; x < c.width; ++x) {
      const int i = x + y * c.width;
      const bool body = bodyAt(double(x)/h, double(y)/h, c.shape, c.angle);
      wall[i] = y == 0 || y == h - 1; solid[i] = body || wall[i];
      if (body && !wall[i]) temperature[i] = c.heat;
      u[i] = solid[i] ? 0 : inlet(y);
      for (int q = 0; q < 9; ++q) f[i*9+q] = equilibrium(q, 1, u[i], 0);
    }
    buildFan();
  }
  double inlet(int y) const {
    if (c.source) return 0;
    if (c.shape != 2) return c.speed;
    const double t = (double(y)/h - .32) / .08;
    return t > 0 && t < 1 ? c.speed * 4 * t * (1 - t) : 0;
  }
  void update(Config config) {
    config.validate();
    if (config.width != c.width || config.shape != c.shape) throw std::runtime_error("Shape and resolution changes require reset.");
    std::vector<unsigned char> mask(n);
    for (int i = 0; i < n; ++i) mask[i] = wall[i] || bodyAt(double(i % c.width)/h, double(i / c.width)/h, config.shape, config.angle);
    // ponytail: changed cells re-equilibrate; use conservative moving boundaries for quantitative pitching-wing loads.
    for (int i = 0; i < n; ++i) {
      if (solid[i] != mask[i]) {
        double r = 0, ux = 0, vy = 0, t = 0; int count = 0;
        for (int j : {i-1, i+1, i-c.width, i+c.width}) if (j >= 0 && j < n && std::abs(j % c.width - i % c.width) <= 1 && !solid[j] && !mask[j]) {
          r += rho[j]; ux += u[j]; vy += v[j]; t += temperature[j]; ++count;
        }
        r = count ? r/count : 1; ux = count ? ux/count : 0; vy = count ? vy/count : 0;
        if (mask[i]) ux = vy = 0;
        rho[i] = r; u[i] = ux; v[i] = vy; temperature[i] = count ? t/count : 20;
        for (int q = 0; q < 9; ++q) f[i*9+q] = next[i*9+q] = equilibrium(q, r, ux, vy);
      }
      if (mask[i]) temperature[i] = wall[i] ? 20 : config.heat;
    }
    c = config; solid.swap(mask); buildFan();
  }
  void step() {
    if (failed) throw std::runtime_error("Solver stopped after instability. Reset required.");
    const int w = c.width;
    const double omega = 1 / (.5 + 3 * c.viscosity);
    for (const auto& [i, strength] : fan) {
      const double ux = u[i] + strength * (fanU - u[i]), vy = v[i] + strength * (fanV - v[i]);
      for (int q = 0; q < 9; ++q) f[i*9+q] += equilibrium(q, rho[i], ux, vy) - equilibrium(q, rho[i], u[i], v[i]);
      u[i] = ux; v[i] = vy;
    }
    // BGK collision, push streaming, halfway no-slip bounce-back.
    for (int i = 0; i < n; ++i) {
      if (solid[i]) continue;
      const int x = i % w, y = i / w, base = i*9;
      for (int q = 0; q < 9; ++q) {
        const double value = f[base+q] + omega * (equilibrium(q, rho[i], u[i], v[i]) - f[base+q]);
        const int nx = x + CX[q], ny = y + CY[q], j = nx + ny*w;
        if (ny < 0 || ny >= h || (nx >= 0 && nx < w && solid[j])) next[base+OPP[q]] = value;
        else if (nx >= 0 && nx < w) next[j*9+q] = value;
      }
    }
    for (int y = 1; y < h - 1; ++y) {
      const int i = y*w, b = i*9;
      if (!solid[i] && c.source) {
        for (int q = 0; q < 9; ++q) next[b+q] = next[b+9+q];
      } else if (!solid[i]) {
        const double ux = inlet(y);
        const double r = (next[b] + next[b+2] + next[b+4] + 2*(next[b+3] + next[b+6] + next[b+7])) / (1-ux);
        next[b+1] = next[b+3] + 2*r*ux/3;
        next[b+5] = next[b+7] + (next[b+4]-next[b+2])/2 + r*ux/6;
        next[b+8] = next[b+6] + (next[b+2]-next[b+4])/2 + r*ux/6;
      }
      const int out = i+w-1;
      if (!solid[out]) for (int q = 0; q < 9; ++q) next[out*9+q] = next[(out-1)*9+q];
    }
    f.swap(next);
    for (int i = 0; i < n; ++i) {
      if (solid[i]) continue;
      double r = 0, ux = 0, vy = 0;
      for (int q = 0; q < 9; ++q) { const double value = f[i*9+q]; r += value; ux += value*CX[q]; vy += value*CY[q]; }
      ux /= r; vy /= r;
      if (!std::isfinite(r+ux+vy) || r < .5 || r > 1.5 || ux*ux+vy*vy > .04) {
        failed = true;
        throw std::runtime_error("Low-Mach stability limit exceeded. Reduce speed or increase viscosity, then reset.");
      }
      rho[i] = r; u[i] = ux; v[i] = vy;
    }
    // ponytail: first-order upwind heat transport is diffusive; upgrade for quantitative heat-transfer studies.
    const double alpha = c.viscosity / .71;
    for (int y = 0; y < h; ++y) for (int x = 0; x < w; ++x) {
      const int i = x+y*w;
      if (solid[i]) { nextTemperature[i] = wall[i] ? 20 : c.heat; continue; }
      if (x == 0) { nextTemperature[i] = 20; continue; }
      const double t = temperature[i], left = temperature[i-1], right = temperature[x == w-1 ? i : i+1], top = temperature[i-w], bottom = temperature[i+w];
      nextTemperature[i] = t - u[i]*(u[i] >= 0 ? t-left : right-t) - v[i]*(v[i] >= 0 ? t-top : bottom-t) + alpha*(left+right+top+bottom-4*t);
    }
    }
    temperature.swap(nextTemperature); ++steps;
  }
  val snapshot() {
    double reference = 0; int count = 0; maxSpeed = 0;
    for (int y = 1; y < h-1; ++y) { const int i = (y+1)*c.width-1; if (!solid[i]) { reference += rho[i]; ++count; } }
    reference /= count ? count : 1;
    for (int i = 0; i < n; ++i) {
      fields[i*4] = u[i]; fields[i*4+1] = v[i]; fields[i*4+2] = solid[i] ? 0 : (rho[i]-reference)/3; fields[i*4+3] = temperature[i];
      maxSpeed = std::max(maxSpeed, std::hypot(u[i], v[i]));
    }
    val result = val::object(); result.set("fields", view("fields")); result.set("steps", steps);
    result.set("maxSpeed", maxSpeed); result.set("reynolds", c.speed*(h/4.)/c.viscosity); return result;
  }
  val view(const std::string& name) {
    if (name == "solid") return val(emscripten::typed_memory_view(solid.size(), solid.data()));
    if (name == "wall") return val(emscripten::typed_memory_view(wall.size(), wall.data()));
    auto* data = name == "f" ? &f : name == "rho" ? &rho : name == "u" ? &u : name == "v" ? &v : name == "temperature" ? &temperature : name == "fields" ? &fields : nullptr;
    if (!data) throw std::runtime_error("Unknown field.");
    return val(emscripten::typed_memory_view(data->size(), data->data()));
  }
  double getSteps() const { return steps; }
};
EMSCRIPTEN_BINDINGS(flow) {
  emscripten::value_object<Config>("Config")
    .field("width", &Config::width).field("shape", &Config::shape).field("source", &Config::source)
    .field("speed", &Config::speed).field("viscosity", &Config::viscosity).field("angle", &Config::angle)
    .field("heat", &Config::heat).field("fanX", &Config::fanX).field("fanY", &Config::fanY).field("direction", &Config::direction);
  emscripten::class_<Solver>("Solver").constructor<Config>()
    .function("step", &Solver::step).function("update", &Solver::update).function("snapshot", &Solver::snapshot)
    .function("view", &Solver::view).function("inlet", &Solver::inlet).property("steps", &Solver::getSteps);
  emscripten::function("equilibrium", &equilibrium);
  emscripten::function("bodyAt", &bodyAt);
}